### DISCUSSION TITLE
How to properly filter a mesh?

### DISCUSSION THREAD

**Discussion by stla** [CGAL/cgal#7169] [Q&A]
Hello,

When I filter a mesh by using successive `mesh.remove_face(fd)` followed by `mesh.collect_garbage()`, I end up with a `mesh` containing unreferenced vertices (i.e. vertices not involved in the faces). Doing `mesh.remove_isolated_vertices()` removes nothing. That's consistent with the following observation: when I do `mesh.degree(vd)` for the vertex descriptor `vd` of such an isolated vertex, that does not return `0`, and `CGAL::faces_around_target(mesh.halfedge(vd), mesh)` returns some faces which do not belong to `mesh`, they belong to the mesh before the filtering.

`Face_filtered_graph` does a better work. There's no isolated vertex after applying this approach. However `CGAL::faces_around_target(mesh.halfedge(vd), mesh)` can return some unexisting face indices (`f42944967295`), and `mesh.degree(vd)` for such a `vd` is wrong since it counts this unexisting face.

An approach which works flawlessly is to completely reconstruct the filtered mesh by starting with an empty mesh and using `add_face` and `add_vertex`.

**Comment by sloriot** [comment]
You should use [`CGAL::Euler::remove_face()`](https://doc.cgal.org/latest/BGL/group__PkgBGLEulerOperations.html#gacfae7ff8e782da55b941e4487e86c738). The function you are using is a low-level one. You can even find the following warning [here](https://doc.cgal.org/latest/Surface_mesh/classCGAL_1_1Surface__mesh.html):
*Functions in this group do not adjust any of connected elements and usually leave the surface mesh in an invalid state.*

**Comment by stla** [comment]
In fact it's nice with `Face_filtered_graph`. The strange face `f42944967295` is nothing but the null face. So I do something like that to count the number of faces involving a given vertex `vd`:

```cpp
    for(face_descriptor fd : CGAL::faces_around_target(mesh.halfedge(vd), mesh)) {
      if(fd != EMesh3::null_face()) {
        // count this face
      }
    }
```
